Enable per-conversation loading states to allow having parallel conversations

allozaur opened the pull request:
Close #16133
Close #16398

Introduces granular loading state management for individual conversations, allowing concurrent message processing which prevents UI lockup and improves user experience when multiple conversations are active.

allozaur commented:
@ggerganov here you go (17e80c5) 😉

ggerganov commented:
Noticed 2 issues:

- The "regenerate" button stops working after I switch to another conv
- The displayed stats do not correspond to the selected conv while another conv is being generated

allozaur committed:
feat: Per-conversation loading states and tracking streaming stats

allozaur committed:
chore: update webui build output

allozaur committed:
refactor: Chat state management

Consolidates loading state management by using a global `isLoading` store synchronized with individual conversation states.

This change ensures proper reactivity and avoids potential race conditions when updating the UI based on the loading status of different conversations. It also improves the accuracy of statistics displayed.

Additionally, slots service methods are updated to use conversation IDs for per-conversation state management, avoiding global state pollution.

allozaur committed:
feat: Adds loading indicator to conversation items

allozaur committed:
chore: update webui build output

allozaur committed:
fix: Fix aborting chat streaming

Improves the chat stream abortion process by ensuring that partial responses are saved before the abort signal is sent.

This avoids a race condition where the onError callback could clear the streaming state before the partial response is saved. Additionally, the stream reading loop and callbacks are now checked for abort signals to prevent further processing after abortion.

allozaur committed:
refactor: Remove redundant comments

allozaur committed:
chore: build webui static output

allozaur committed:
refactor: Cleanup

allozaur committed:
chore: update webui build output

ServeurpersoCom commented:
Rebased on https://github.com/allozaur/llama.cpp/tree/16133-parallel-streaming + 1 commit https://github.com/ggml-org/llama.cpp/commit/dfd3ab8353078bf4da9272df0d17db8b59337d97 (for model JSON payload)

--parallel 5 UI monkey test:

https://github.com/user-attachments/assets/62a57a2c-6718-4cf1-8003-01b789e36981

Nothing to report, except that it takes a short moment for the slot to be released.

allozaur commented:
Thanks, @ServeurpersoCom for testing this on your end. @ggerganov it would be great to hear an update from you on this as well. LMK!

allozaur committed:
Merge remote-tracking branch 'origin/master' into 16133-parallel-streaming

allozaur committed:
chore: update webui build output

ggerganov commented:
Works great. One problem - using "Regenerate" button does not trigger the "processing" indicator in the side panel. It is only triggered when submitting a new message.

allozaur committed:
fix: Conversation loading indicator for regenerating messages

allozaur committed:
chore: update webui static build

allozaur commented:
@ggerganov just pushed an update addressing the loading indicator for regenerated messages.

allozaur committed:
feat: Improve configuration

allozaur commented:
There's some issues with `npm`'s availability right now so the CI can't finish the E2E tests which are using `npx http-server`.

<img width="892" height="378" alt="Zrzut ekranu 2025-10-20 o 11 45 54" src="https://github.com/user-attachments/assets/ec6021fe-7ade-4098-b19e-07c62c40609a" />

allozaur committed:
feat: Install `http-server` as dev dependency to not need to rely on `npx` in CI
